File: 496-next-greater-element-i/next-greater-element-i.java
import java.util.*;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {

        HashMap<Integer, Integer> map = new HashMap<>();
        Stack<Integer> stack = new Stack<>();

        // Step 1: process nums2
        for(int x : nums2) {

            while(!stack.isEmpty() && stack.peek() < x) {
                map.put(stack.peek(), x);
                stack.pop();
            }

            stack.push(x);
        }

        // Step 2: remaining → -1
        while(!stack.isEmpty()) {
            map.put(stack.peek(), -1);
            stack.pop();
        }

        // Step 3: build answer
        int[] ans = new int[nums1.length];

        for(int i = 0; i < nums1.length; i++) {
            ans[i] = map.get(nums1[i]);
        }

        return ans;
    }
}

File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {

        unordered_map<int, int> mp; // element → next greater
        stack<int> st;

        // Step 1: process nums2
        for(int x : nums2) {

            while(!st.empty() && st.top() < x) {
                mp[st.top()] = x;
                st.pop();
            }

            st.push(x);
        }

        // Step 2: remaining elements → -1
        while(!st.empty()) {
            mp[st.top()] = -1;
            st.pop();
        }

        // Step 3: build answer
        vector<int> ans;

        for(int x : nums1) {
            ans.push_back(mp[x]);
        }

        return ans;
    }
};